src/wasm/wasm-s-parser.cpp

```diff
@@ -3291,11 +3291,17 @@ void SExpressionWasmBuilder::parseElem(Element& s, Table* table) {
     return;
   }
 
-  if (s[i]->isStr() && s[i]->str() == FUNC) {
-    auto segment = std::make_unique<ElementSegment>();
-    segment->setName(name, hasExplicitName);
-    parseElemFinish(s, segment, i + 1);
-    return;
+  if (s[i]->isStr()) {
+    auto makePassiveSegment = [&](bool usesExpressions) {
+      auto segment = std::make_unique<ElementSegment>();
+      segment->setName(name, hasExplicitName);
+      parseElemFinish(s, segment, i + 1, usesExpressions);
+    };
+    if (s[i]->str() == FUNC) {
+      return makePassiveSegment(false);
+    } else if (s[i]->str() == FUNCREF) {
+      return makePassiveSegment(true);
+    }
```

Comment:
It looks like this could be simplified by just calculating a bool `usesExpressions` first, then running what is now the body of the lambda.

Reply:
Done. Although it required an additional `isPassive` variable as well. I added a comment for the `else` case.